DiscordBot/commands/mud/attack.js

```diff
@@ -0,0 +1,128 @@
+    }
+  }
+
+  checkRoom(message, data, args, player) {
+    const room = JSON.parse(data.body).Item;
+    const enemy = this.cleanArgs(args).object;
+    //look for the npc
+    if(room.npcs[enemy]) {
+      db.getItem(room.npcs[enemy], 'entities', (data) => this.checkHostile(message, player, data));
+    } else {
+      message.reply('glares with murderous intent towards no one in particular');
+    }
+  }
+
+  checkHostile(message, player, data) {
+    const enemy = JSON.parse(data.body).Item;
+    try {
+      //this is a fancy way of making sure enemy is defined
+      //otherwise trying to access a key of an unef object throws a big error
+      if(enemy && enemy.hostile) {
+        if(enemy.aggro === 'nobody' || player.id) {
+          this.combatLoop(message, player, enemy);
+        } else {
+          message.reply('attempts to encroach on existing combat, and fails.');
+        }
+      } else {
+        message.reply(`glares with murderous intent towards ${enemy.name}.`);
+      }
+    } catch (err) {
+      console.log(`${err}\nThis indicates an enemy not respawning correctly or quick enough`);
+      message.reply('tries to attack the air');
+    }
+  }
+
+  cleanArgs(args) {
+    // ignore the argument's capitalization
+    args.object = args.object.toLowerCase();
+    return args;
+  }
+
+  combatLoop(message, player, enemy) {
+    db.updateItem(player.id, ['busy'], [true], 'players', ()=>{});
+    db.updateItem(enemy.id, ['aggro'], [player.id], 'entities', () => {});
+    while (player.health > 0 && enemy.health > 0) {
+      //calculate player damage on enemy and update value
+      let damage = player.strength - enemy.defense;
+      if (damage > 0) {
+        enemy.health = enemy.health - damage;
+        message.reply(`hit ${enemy.name} for ${damage.toString()} damage.`);
+      } else {
+        message.reply(`swung at the ${enemy.name} and missed.`);
+      }
+      //prevents enemy attacking if dead
+      if(enemy.health <= 0) {
+        break;
+      }
+      //calculate enemy damage on agro target and update value
+      damage = enemy.strength - player.defense; //for the following lines replace player with agro target
+      if (damage > 0) {
+        player.health = player.health - damage;
+        message.reply(`was hit by ${enemy.name} for ${damage.toString()} damage.`);
+      } else {
+        message.reply(`${enemy.name} swung at ${player.name} and missed.`);
+      }
+    }
+    console.log('updating player state');
+    db.updateItem(player.id, ['health', 'busy'], [player.health, false], 'players', ()=>{});
+    if(enemy.health <= 0) {
+      message.reply(`defeated the ${enemy.name}.`);
+      //TODO: loot roll here
+      /* TODO: move this delete to the end of the loot roll so 
+      we don't delete the enemy before distributing their loot */
+      db.deleteItem(enemy.id, 'entities', ()=>{});
+    } else {
```

Comment:
If the enemy health isn't less than or equal to 0, and we've exited the loop that checked whether or not the player and the enemy's health was above 0, then isn't it implied that the player's health is less than or equal to 0? If so, we should move the player death code into this else statement.

Reply:
oh hey yeah! nice catch.